### PR TITLE
Elasticstore

### PR DESCRIPTION
**Pull request by pdwood**
Currently uses elasticstore 6.7.1

To enable, uncomment the definition of `knowledge_elasticsearch` in config_defaults.py

### CODE REVIEW

**Comment by gordom6** [comment]
Setup instructions:
- ElasticSearch 6.7.1 on localhost
- Uncomment knowledge_elasticsearch in config_defaults.py
- Run tests as usual

**Comment by gordom6** [comment]
Various tests are failing currently because of Not implemented assertions in the elasticstore.py.

**Comment by gordom6** [comment]
@pdwood TestCase needs __init__ to be extended to set knowledge_elasticsearch in the config.Test if an environment variable is set ("HAVE_ELASTICSEARCH") Once all of the tests pass I will modify the CircleCI configuration to run the tests once without this environment variable set (to test blazegraph) and once with.

File: whyis/database/elasticstore.py [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+    __namespace = {}
+
+    def __init__(self, configuration=None, identifier=None):
+        self.__open = False
+        self.__identifier = identifier
+        super(ElasticSearchStore, self).__init__(configuration)
+
+    def __get_identifier(self):
+        return self.__identifier
+    identifier = property(__get_identifier)
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: `@property`

File: whyis/database/elasticstore.py [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+    def open(self, url, create=True):
+        """\
+        Opens the store.
+
+        The given url should include the elasticsearch index that will be used (for example, "http://example.com:9200/some_index").
+        If create is True, the index will be created if it does not already exist.
+        """
+        self.url = url
+        self.session = requests.Session()
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: Private

File: whyis/database/elasticstore.py [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+    formula_aware = False
+    transaction_aware = False
+    graph_aware = True
+    __prefix = {}
+    __namespace = {}
+
+    def __init__(self, configuration=None, identifier=None):
+        self.__open = False
+        self.__identifier = identifier
+        super(ElasticSearchStore, self).__init__(configuration)
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: All members should be set to dummy values in __init__, such as self.__session = None

File: whyis/database/elasticstore.py [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+        If create is True, the index will be created if it does not already exist.
+        """
+        self.url = url
+        self.session = requests.Session()
+        self.session.headers.update({"Content-Type":"application/json"})
+        status = self.session.get(url)
+        if create and status.status_code == 404:
+            r = self.session.put(self.url,data=elastic_index_settings)
+            if r.status_code != 201:
+                print(r.status_code)
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: logging

**Comment by pdwood** [comment]
> 
> 
> Various tests are failing currently because of Not implemented assertions in the elasticstore.py.

Specifically, these tests call the `contexts` method of ElasticSearchStore.

File: whyis/database/elasticstore.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+                print(r.content)
+        self.__open = True
+
+        return VALID_STORE
+
+    def sync(self):
+        if self.__open:
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: ?

Comment (gordom6) [reply whyis/database/elasticstore.py]: refresh?

File: whyis/database/elasticstore.py [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+                }
+              },
+              "path": "graphs.@graph",
+              "score_mode": "avg"
+            }
+          }
+        }
+
+        print("query is", query)
+        self.session.post(self.url+"/_refresh") # Ensure store is up-to-date before reading
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: Duplicate code, factor out

File: whyis/database/elasticstore.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+            match_all = False
+            query['bool']['must'].append( { "term" : {
+                 "graphs.@id" : str(context)
+                } } )
+
+        if match_all:
+            query = {'match_all': {}}
+
+        #print json.dumps(query)
+        json_response = self.elastic_query(query)
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: Private

File: whyis/database/elasticstore.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+        self.__namespace[prefix] = namespace
+
+    def namespace(self, prefix):
+        return self.__namespace.get(prefix, None)
+
+    def prefix(self, namespace):
+        return self.__prefix.get(namespace, None)
+
+    def namespaces(self):
+        for prefix, namespace in iteritems(self.__namespace):
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: yield from

File: whyis/database/elasticstore.py [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+
+    def add_graph(self, graph, refresh=False):
+        self.addN(((s,p,o,graph.identifier) for s,p,o in graph.triples((None, None, None))), refresh=refresh)
+
+    def remove_graph(self, graph):
+        if isinstance(graph, Graph):
+            graph = graph.identifier
+        response = self.session.delete(self.url+"/nanopublication/"+graph.split('/')[-1])
+
+def json_ld_triples(json_response, subject=None, predicate=None, object=None, context=None):
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: private @staticmethod

File: whyis/database/elasticstore.py [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+                    for obj in resource[pred]:
+                        o = None
+                        if '@id' in obj:
+                            id = obj['@id']
+                            if id.startswith('_:'):
+                                o = rdflib.BNode(id[2:])
+                            else:
+                                o = rdflib.URIRef(id)
+                        else:
+                            args = {}
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: kwds or kwargs

File: whyis/database/elasticstore.py [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+                            if '@type' in obj:
+                                args['datatype'] = rdflib.URIRef(obj['@type'])
+                            if '@lang' in obj:
+                                args['lang'] = obj['@lang']
+                            o = rdflib.Literal(obj['@value'],**args)
+                        if object is not None and o != object:
+                            continue
+                        yield resourceid, rdflib.URIRef(pred), o, graphid
+
+def evalBGP(ctx, part=None, bgp=None):
```
Comment (gordom6) [inline whyis/database/elasticstore.py]: Duplicated?

File: main.py [hunk 12]
```diff
@@ -294,10 +294,12 @@ def this(self):
 
         def description(self):
             if self._description is None:
-#                try:
-                result = Graph()
-#                try:
-                for quad in self._graph.query('''
+                result = self._graph.store.subgraph({ "term" : { "graphs.@graph.@id" : str(self.identifier) } })
```
[the remark this answers is not in the document]
Comment (pdwood) [reply main.py]: This does indeed break blazegraph. I'll set it up to check if elasticstore is in use here.

File: whyis/filters.py [hunk 13]
```diff
@@ -79,6 +79,8 @@ def query_filter(query, graph=app.db, prefixes=None, values=None):
         params = { 'initNs': namespaces}
         if values is not None:
             params['initBindings'] = values
+        print(query)
```
[the remark this answers is not in the document]
Comment (pdwood) [reply whyis/filters.py]: This was... probably not supposed to be committed.

File: whyis/database/elastic_index_settings.json [hunk 14]
```diff
@@ -1,5 +1,6 @@
 {
     "settings": {
+        "index.mapping.ignore_malformed": true
```
[the remark this answers is not in the document]
Comment (pdwood) [reply whyis/database/elastic_index_settings.json]: All tests pass when this change is removed. Reverted.

File: whyis/database/elasticstore.py (hunk 11 above)
Comment (pdwood) [reply whyis/database/elasticstore.py]: This code was not used, and as such was removed.

File: whyis/database/elasticstore.py (hunk 10 above)
Comment (pdwood) [reply whyis/database/elasticstore.py]: Changed to `kwargs`

File: whyis/database/elasticstore.py (hunk 8 above)
Comment (pdwood) [reply whyis/database/elasticstore.py]: Changed to `yield from`

File: whyis/database/elasticstore.py (hunk 6 above)
Comment (pdwood) [reply whyis/database/elasticstore.py]: What are you referring to here? The call to `/_refresh`?

File: whyis/database/elasticstore.py (hunk 7 above)
Comment (jimmccusker) [reply whyis/database/elasticstore.py]: What should be private, elastic_query?

Comment (minorg) [reply whyis/database/elasticstore.py]: Yes, the store should encapsulate access to ES.

File: whyis/database/elasticstore.py (hunk 6 above)
Comment (minorg) [reply whyis/database/elasticstore.py]: Yes.

**Comment by jimmccusker** [comment]
I disagree that ES should be entirely encapsulated. It should be possible to retrieve nanopublications that match ES queries directly itself. That would require a commitment to elasticsearch, but the whole point of this is that there are situations where that would be desirable and much faster than even a conventional SPARQL query.

**Pull request closed without merging**